[Doc] Update llvm_sparse_runtime.md (#7215)
taichi-dev/taichi: docs/design/llvm_sparse_runtime.md

# SNode

There are four types of SNode in Taichi: `dense`, `bitmasked`, `pointer` and `dynamic`. SNodes of `dynamic` or `pointer` types are spatially sparse in nature: cells in them are not neccessarily stored contiguously. So keep in mind that when we talk about sparse SNodes hereafter, we are referring to `dynamic` and `pointer` SNodes.

For each SNode type of the four listed above, it comes with a `XMeta` struct derived from [`StructMeta`](https://github.com/taichi-dev/taichi/blob/2cdc58078ecd2aef2cde608f07325108c5b3d5a5/taichi/runtime/llvm/runtime.cpp#L365-L383). `StructMeta` has the following properties:

* `snode_id`: SNode ID.
* `i32 X_get_num_elements(Ptr meta, Ptr node)`: Returns the capacity this SNode can hold. Note that it is *not* the current number of active cells, but the maximum.
* `void X_activate(Ptr meta, Ptr node, int i)`: Activates cell `i`.
* `i32 X_is_active(Ptr meta, Ptr node, int i)`: Checks whether cell `i` is active.

> * `Ptr X_lookup_element(Ptr meta, Ptr node, int i)`: Returns the pointer to cell `i`. This can be `nullptr` for sparse SNodes.

Reviewer: Not sure what "this" refers to. -> The result can be ...

Here, `Ptr` is an alias for `uint8_t*`, and `i32` for `int32_t`. As the name suggests, `meta` points to the corresponding meta struct, while `node` points to the SNode instance.

For sparse SNodes, they also implement an additional API:

* `void X_deactivate(Ptr meta, Ptr node, int i)`: Deactivates cell `i`.

> However, this additional API is likely subject to change, so that all SNodes can share the same set of APIs in the future.

Reviewer: ```suggestion
However, this additional API is subject to change to make all SNodes share the same set of APIs in the future.
```

> `pointer` is usually the go-to sparse SNode you should consider. It allocates memory only when a cell is actually activated, and recycles to a memory pool once it is deactivated. This saves the memory resource in large-scale grid computation. You can mentally view it as an `std::array<Cell*, N>`.

Reviewer: ```suggestion
`pointer` is usually the go-to sparse SNode. It allocates memory only when a cell is activated, and recycles the memory to a memory pool once the cell is deactivated. In this way, it saves the memory resource in large-scale grid computation. You can regard it as an `std::array<Cell*, N>`.
```

> Upon initialization, Taichi preallocates a chunk of memory space named `ambient_elements`, which is shared across all inactive sparse SNodes. Therefore dereferencing an inactive sparse SNode will generate the default value (usually zero) stored in `ambient_elements`.

Reviewer: ```suggestion
Upon initialization, Taichi preallocates a chunk of memory space named `ambient_elements`, which is shared across all inactive sparse SNodes. Therefore, dereferencing an inactive sparse SNode generates the default value (usually zero) stored in `ambient_elements`.
```

```

1. Uses the current length `n` as the index (`i`) to store `data`.
2. `chunk_start` tracks the starting index of a given chunk, and always starts at `0`. `p_chunk_ptr` is initialized to the pointer to the first chunk.
3. Inside the `while` loop, checks if the given chunk slot is empty first, and allocates a new chunk if so.
4. Compares if the determined index `i` falls within the current chunk.
   1. If so, stores `data` into the corresponding slot in this chunk. Note that the first `sizeof(Ptr)` bytes are skipped: they are reserved to store the address of the next chunk.
   2. Otherwise, jumps to the next chunk.

# Runtime

> Runtime for the LLVM backends is in https://github.com/taichi-dev/taichi/blob/master/taichi/runtime/llvm/runtime_module/runtime.cpp. Note that this file is *NOT* linked into Taichi's core C++ library. Instead, it is compiled into a LLVM byte code file (`.bc`). Upon starting Taichi, the `.bc` file is loaded back into the memory, de-serialized into an `llvm::Module`, and linked together with the JIT compiled Taichi kernels. This design has the advantage that the runtime code can be written once and shared between backends using LLVM, such as CPU and CUDA. In addition, the sparse runtime can be implemented in any language with enough abstraction (e.g., C++) instead of raw LLVM IR.

Reviewer: ```suggestion
Runtime for the LLVM backends is in https://github.com/taichi-dev/taichi/blob/master/taichi/runtime/llvm/runtime_module/runtime.cpp. Note that this file is *NOT* linked into Taichi's core C++ library. Instead, it is compiled into an LLVM byte code file (`.bc`). Upon starting Taichi, the `.bc` file is loaded back into the memory, de-serialized into an `llvm::Module`, and linked together with the JIT compiled Taichi kernels. This design has the advantage that the runtime code can be written once and shared between backends using LLVM, such as CPU and CUDA. In addition, the sparse runtime can be implemented in any language with enough abstraction (e.g., C++) instead of raw LLVM IR.
```

The core data structure of this runtime is [`LLVMRuntime`](https://github.com/taichi-dev/taichi/blob/172cab8a57fcfc2d766fe2b7cd40af669dadf326/taichi/runtime/llvm/runtime.cpp#L543), which holds a handful of data:

* All the root SNodes info (`roots` and `root_mem_sizes`).
* SNode memory allocators.
* Random states (`rand_states`) used by `ti.random`.
* Print and error message buffers.
* ...

> SNode memory allocator is the bedrock of sparse SNodes, We will explain how it's implemented in the next sections.

Reviewer: ```suggestion
The SNode memory allocator is the bedrock of sparse SNodes. The following sections explain how it is implemented. 
```

}
```

1. Reads the (possible) index of the next free cell.
2. If running out of the indices in the free list, allocate a new chunk from `data_list`.
3. Otherwise, re-use the index from `free_list`.
4. Either way, index `l` points to a memory slot in `data_list`. Returns that slot.

> This finishes our analysis of the `allocate` procedure. For the recycle procedure, the code [`recycle()`](https://github.com/taichi-dev/taichi/blob/172cab8a57fcfc2d766fe2b7cd40af669dadf326/taichi/runtime/llvm/runtime.cpp#L672-L675) should be straightforward to understand.

Reviewer: ```suggestion
So far, we have analyzed the `allocate` procedure. For the recycle procedure, the code [`recycle()`](https://github.com/taichi-dev/taichi/blob/172cab8a57fcfc2d766fe2b7cd40af669dadf326/taichi/runtime/llvm/runtime.cpp#L672-L675) is self-explanatory.
```

}
```

> It is important to understand that this stage must be run on a single thread. This is to avoid any kind of data race on `recycled_list`. During GC, `recycled_list` must be cleared, with all its contents being transferred into `free_list`. If this was done in the third stage (which is run in parallel), it would be quite difficult to coordinate the event sequence of clearing and transferring among the GPU threads. As a result, this serial stage is created in order to store the number of elements of `recycled_list` into `recycle_list_size_backup` in advance (1), then clears the list (2). Fortunately, the computation task here is light and it doesn't take too much time even running in serial.

Reviewer: ```suggestion
It is important to understand that this stage must run on a single thread to avoid data races on `recycled_list`. During GC, `recycled_list` must be cleared, with all of its contents transferred into `free_list`. If this were done in the third stage (which runs in parallel), it would be difficult to coordinate the event sequence of clearing and transferring among the GPU threads. As a result, this serial stage is created to store the number of elements of `recycled_list` into `recycle_list_size_backup` in advance (1) and then clear the list (2). Fortunately, the computation task here is light and does not take too much time even running in serial.
```
